README.md
<!--
see more recent:

   https://computing.llnl.gov/projects/odepack

might have test cases:

   https://people.sc.fsu.edu/~jburkardt/f77_src/odepack/odepack.html

markdown references

   https://michelf.ca/projects/php-markdown/extra
-->

![ODEPACK](docs/images/odepack.png)
============

# module ODEPACK

This is a WIP(Work In Progress) to evaluate the effort required and
feasibility of updating older Fortran code from the netlib repository
using a combination of the commercial plusFORT/spag software and GNU
utilities as well as conventional manual refactoring.

Many thanks to plusFORT for making an evaluation copy available for
several months to the Fortran Community. The plusFORT tools were crucial
to this study and unmatched in my experience for removing deprecated
syntax from pre-f2003 code.

This began with [ODEPACK](https://computing.llnl.gov/projects/odepack)

the ODEPACK package has been selected as a significant code that is well
documented and structured for its vintage, and available on the netlib
site and covered by a public domain license.

#### preliminary targets for phase I:

 + remove obsolescent syntax (trying plusFORT and spag)
 + able to build using fpm(1)  (the Fortran Package Manager)
   - in debug mode (ongoing)
   - with ifort, gfortran, nvfortran
 + text viewable in ford(1) and extractable as markup that can run through pandoc(1) (ongoing)
 + available on github (or equivalent)
 + no common blocks
 + no equivalences
 + build as a (single?) module M_odepack (ongoing)
 + complete set of unit tests (tests currently only use the original examples)

The originally bundled subset of BLAS/LAPACK routines are
being included in the module.  In production, this might not be done in
order to be able to easily call external optimized versions.

The biggest hinderance is some storage used for both INTEGER and DOUBLE
PRECISION values.

One take-away is how critical unit tests are to enable rapid development
(which so far this package does not have)

The initial pass was done just using the original sample programs as
unit tests. This may have allowed for introduction of errors as this
is a WIP but the original samples run with the same output as the original.

plusFORT was invaluable and reduced the effort by an estimated 85
percent. The results have been encouraging enough to inspire completing
the transformation.

### Phase II ?

Some of the goals of phase I are incomplete, but The results of this first
pass were significant enough that this project will hopefully continue.

A complete unit test suite is required to allow development to proceed
rapidily. Contributions, especially from current ODEPACK users are
particularly welcome.

Another major issue is the remaining non-standard code.  Non-standard (but
at the time de-facto fortran standard) such as equivalencing different
types, creating scratch space that is used as different numeric types,
and treating scalars as arrays and vice-versa as well as passing the same
arrays or values multiple times are the most time-consuming usages to
correct to standard-conforming, particularly since spag(1) had already
done an excellent job with updating the pre-f2003 code. spag(1) is not
(currently?) sufficient by itself to automate the additional refactoring
desired, which includes using post-f95 features and code restructuring,
so the remaining work requires manual recoding.

The type-mismatch issues have not been eliminated enough to include all
the routines in the module, so  those in the files "M_da1/dprep.inc"
"M_da1/dainvgs.inc" "M_da1/dprepi.inc" "M_da1/dstodi.inc" and
"M_da1_/dstode.inc" still require being built without an interface
definition.

### building with fpm(1)

This version of ODEPACK already builds with an included make(1) file
and as an fpm(1) package with the current options:
```bash
 fpm run                     --compiler nvfortran --example '*'
 fpm run --profile release   --compiler ifort     --example '*'
 fpm run --flag -std=legacy  --compiler gfortran  --example '*'
```
### building with make(1)
```bash

cd src
# gfortran
make clean
make gfortran
make run
make test

# ifort
make clean
make ifort
make run
make test

# nvfortran
make clean
make nvfortran
make run
make test
```
### To rebuild the documentation with ford(1), use

```bash
 ford ford.md
```

The code is far more readable after having been refactored by
a combination of using spag(1) from the plusFORT package and manual
editing, and is believed as useable as the original.

There are a few notes in src/M_odepack.f90 concerning continuing issues.

Current users of ODEPACK are encouraged to try this version and provide
feedback.

Hopefully as a community we can complete creating a new maintained
production-quality version of this venerable and still-valuable package.

### Documentation

The ongoing API documentation for the current `master` branch can
be found [here](https://jacobwilliams.github.io/odepack/).
This is generated by processing the source files with
[FORD](https://github.com/Fortran-FOSS-Programmers/ford).

#### Solvers for explicitly given systems.
Links to the solver-specific documenation for the main procedures (as described below):

   + [dlsoda](https://jacobwilliams.github.io/odepack/proc/dlsoda.html)
   + [dlsodar](https://jacobwilliams.github.io/odepack/proc/dlsodar.html)
   + [dlsode](https://jacobwilliams.github.io/odepack/proc/dlsode.html)
   + [dlsodes](https://jacobwilliams.github.io/odepack/proc/dlsodes.html)
   + [dlsodi](https://jacobwilliams.github.io/odepack/proc/dlsodi.html)
   + [dlsodis](https://jacobwilliams.github.io/odepack/proc/dlsodis.html)
   + [dlsodkr](https://jacobwilliams.github.io/odepack/proc/dlsodkr.html)
   + [dlsodpk](https://jacobwilliams.github.io/odepack/proc/dlsodpk.html)
   + [dlsoibt](https://jacobwilliams.github.io/odepack/proc/dlsoibt.html)

Here is an overview primarily from the original documentation  ...

--------------------------------------------------------------------------------

## Brief Description of ODEPACK - A Systematized Collection of ODE Solvers (Double Precision Version)

```text
Alan C. Hindmarsh
Center for Applied Scientific Computing, L-561
Lawrence Livermore National Laboratory
Livermore, CA 94551, U.S.A.

20 June 2001
```

Work performed under the auspices of the U.S. Department of Energy
by the Lawrence Livermore National Laboratory under contract
No. W-7405-Eng-48, and supported (formerly) by the DOE Office of
Energy Research, Applied Mathematical Sciences Research Program.

---------------------------------------------------------------------------

   ODEPACK is a collection of Fortran solvers for the initial value
problem for ordinary differential equation systems.  It consists of nine
solvers, namely a basic solver called LSODE and eight variants of it --
LSODES, LSODA, LSODAR, LSODPK, LSODKR, LSODI, LSOIBT, and LSODIS.
The collection is suitable for both stiff and nonstiff systems.  It
includes solvers for systems given in explicit form, dy/dt = f(t,y),
and also solvers for systems given in linearly implicit form,
A(t,y) dy/dt = g(t,y).  Two of the solvers use general sparse matrix
solvers for the linear systems that arise.  Two others use iterative
(preconditioned Krylov) methods instead of direct methods for these
linear systems.  The most recent addition is LSODIS, which solves
implicit problems with general sparse treatment of all matrices involved.

   The ODEPACK solvers are written in standard Fortran 77, with a few
exceptions, and with minimal machine dependencies.  There are separate
double and single precision versions of ODEPACK.  The actual solver
names are those given above with a prefix of D- or S- for the double
or single precision version, respectively, i.e. DLSODE/SLSODE, etc.
Each solver consists of a main driver subroutine having the same name
as the solver and some number of subordinate routines.  For each
solver, there is also a demonstration program, which solves one or two
simple problems in a somewhat self-checking manner.

   Recently, the ODEPACK solvers were upgraded to improve their
portability in numerous ways.  Among the improvements are (a) renaming
of routines and Common blocks to distinguish double and single
precision versions, (b) use of generic intrinsic function names, (c)
elimination of the Block Data subprogram, (d) use of a portable
routine to set the unit roundoff, and (e) passing of quoted strings to
the error message handler.  In addition, the prologue and internal
comments were reformatted, and use mixed upper/lower case.  Numerous
minor corrections and improvements were also made.

   The above upgrade operations were applied to LSODE earlier than they
were to the rest of ODEPACK, and the two upgrades were done somewhat
independently.  As a result, some differences will be apparent in the
source files of LSODE and the other solvers -- primarily in the
formatting of the comment line prologue of the main driver routine.
In Subroutines DLSODE/SLSODE and their subordinate routines, the
prologue was written in "SLATEC format", while for the other solvers a
more relaxed style was used.  The differences are entirely cosmetic,
however, and do not affect performance.

   Documentation on the usage of each solver is provided in the
initial block of comment lines in the source file, which (in most
cases) includes a simple example.  A demonstration program (in
separate double/single precision versions) is also available.

   What follows is a summary of the capabilities of ODEPACK, comments
about usage documentation, and notes about installing the collection.
For additional documentation on ODEPACK, see also the papers [1], [2]
(for LSODE), and [3] (for LSODPK and LSODKR), and in the references
cited there.  (However, the document [2] does not reflect the upgrade
operations described above.)

### References:

1.  A. C. Hindmarsh, "[ODEPACK, A Systematized Collection of ODE Solvers](https://computation.llnl.gov/casc/nsde/pubs/u88007.pdf),"
     in Scientific Computing, R. S. Stepleman et al. (eds.), North-Holland,
     Amsterdam, 1983 (vol. 1 of IMACS Transactions on Scientific Computation),
     pp. 55-64.

2.  K. Radhakrishnan and A. C. Hindmarsh, "[Description and Use of LSODE,
     the Livermore Solver for Ordinary Differential Equations](https://computation.llnl.gov/casc/nsde/pubs/u113855.pdf)," LLNL
     report UCRL-ID-113855, December 1993.

3.  P. N. Brown and A. C. Hindmarsh, "[Reduced Storage Matrix Methods
     in Stiff ODE Systems](https://computation.llnl.gov/casc/nsde/pubs/u95088.pdf)," J. Appl. Math. & Comp., 31 (1989), pp.40-91.

---------------------------------------------------------------------------

### I. Summary of the ODEPACK Solvers

#### A. Solvers for explicitly given systems.

For each of the following solvers, it is assumed that the ODEs are
given explicitly, so that the system can be written in the form
dy/dt = f(t,y), where y is the vector of dependent variables, and t is
the independent variable.

1. LSODE (Livermore Solver for Ordinary Differential Equations) is the
   basic solver of the collection.  It solves stiff and nonstiff systems
   of the form dy/dt = f.  In the stiff case, it treats the Jacobian
   matrix df/dy as either a dense (full) or a banded matrix, and as either
   user-supplied or internally approximated by difference quotients.
   It uses Adams methods (predictor-corrector) in the nonstiff case,
   and Backward Differentiation Formula (BDF) methods (the Gear methods)
   in the stiff case.  The linear systems that arise are solved by direct
   methods (LU factor/solve).  LSODE supersedes the older GEAR and GEARB
   packages, and reflects a complete redesign of the user interface
   and internal organization, with some algorithmic improvements.

2. LSODES, written jointly with A. H. Sherman, solves systems dy/dt = f
   and in the stiff case treats the Jacobian matrix in general sparse
   form.  It determines the sparsity structure on its own, or optionally
   accepts this information from the user.  It then uses parts of the
   Yale Sparse Matrix Package (YSMP) to solve the linear systems that
   arise, by a sparse (direct) LU factorization/backsolve method.
   LSODES supersedes, and improves upon, the older GEARS package.

3. LSODA, written jointly with L. R. Petzold, solves systems dy/dt = f
   with a dense or banded Jacobian when the problem is stiff, but it
   automatically selects between nonstiff (Adams) and stiff (BDF)
   methods.  It uses the nonstiff method initially, and dynamically
   monitors data in order to decide which method to use.

4. LSODAR, also written jointly with L. R. Petzold, is a variant of
   LSODA with a rootfinding capability added.  Thus it solves problems
   dy/dt = f with dense or banded Jacobian and automatic method
   selection, and at the same time, it finds the roots of any of a
   set of given functions of the form g(t,y).  This is often useful
   for finding stop conditions, or for finding points at which a switch
   is to be made in the function f.

5. LSODPK, written jointly with Peter N. Brown, is a variant of LSODE
   in which the direct solvers for the linear systems have been replaced
   by a selection of four preconditioned Krylov (iterative) solvers.
   The user must supply a pair of routine to evaluate, preprocess, and
   solve the (left and/or right) preconditioner matrices.  LSODPK also
   includes an option for a user-supplied linear system solver to be used
   without Krylov iteration.

6. LSODKR is a variant of LSODPK with the addition of the same
   rootfinding capability as in LSODAR, and also of automatic switching
   between functional and Newton iteration.  The nonlinear iteration
   method-switching differs from the method-switching in LSODA and LSODAR,
   but provides similar savings by using the cheaper method in the non-stiff
   regions of the problem.  LSODKR also improves on the Krylov methods in
   LSODPK by offering the option to save and reuse the approximate Jacobian
   data underlying the preconditioner.


#### B. Solvers for linearly implicit systems.

The following solvers treat systems in the linearly implicit form
A(t,y) dy/dt = g(t,y), A = a square matrix, i.e. with the derivative
dy/dt implicit, but linearly so.  These solvers allow A to be
singular, in which case the system is a differential-algebraic
equation (DAE) system.  In that case, the user must be very careful
to supply a well-posed problem with consistent initial conditions.

7. LSODI, written jointly with J. F. Painter, solves linearly implicit
   systems in which the matrices involved (A, dg/dy, and d(A dy/dt)/dy)
   are all assumed to be either dense or banded.  LSODI supersedes the
   older GEARIB solver and improves upon it in numerous ways.

8. LSOIBT, written jointly with C. S. Kenney, solves linearly implicit
   systems in which the matrices involved are all assumed to be
   block-tridiagonal.  Linear systems are solved by the LU method.

9. LSODIS, written jointly with S. Balsdon, solves linearly implicit
   systems in which the matrices involved are all assumed to be sparse.
   Like LSODES, LSODIS either determines the sparsity structure or
   accepts it from the user, and uses parts of the Yale Sparse Matrix
   Package to solve the linear systems that arise, by a direct method.

---------------------------------------------------------------------------

### II. Usage Documentation

   Each of the solvers in the ODEPACK collection is headed by a
user-callable driver subroutine, with the same name as the solver
(DLSODE, etc.).  The call sequence of the driver routine includes the
names of one or more user-supplied subroutines that define the ODE
system, and various other problem and solution parameters.  Complete
user documentation is given in the initial block of comment lines
(the prologue) in the driver routine.  In each case, this prologue is
organized as follows:

 * Summary of Usage (short, for standard modes of use)
 * Example Problem, with code and output (except for LSODPK and LSODKR)
 * Full Description of User Interface, further divided as follows:
      1. Call sequence description (including optional inputs/outputs)
      2. Optionally callable routines
      3. Descriptions of internal Common blocks
      4. Optionally user-replaceable routines
 * Revision History, showing date written and dates of revisions
 * Other Routines, a list of all subordinate routines for the solver

First-time users should read only the Summary of Usage and look at the
the Example Problem (or demonstration program), then later refer to the
Full Description if and when more details or nonstandard options are needed.

---------------------------------------------------------------------------

### III. Installation Notes

##### Use of supplied matrix procedures

THe src/M_matrix/ directory  includes modified versions of routines from the
LINPACK and BLAS collections that are needed by the solvers (and by two
of the demonstration programs), for the solution of dense and banded
linear systems and associated basic linear algebra operations.
These routine are:

       _From LINPACK_ :  DGEFA, DGESL, DGBFA, DGBSL
       _From the BLAS_: DAXPY, DCOPY, DDOT, DSCAL, DNRM2, IDAMAX

If your computer system already has these routines, and especially if it
has machine-optimized versions, the copies provided in the M_module
module should probably not be called by your program if high performance
is required.

##### The message package

The second auxiliary source file directory M_da1/ includes a set of five
routines -- XERRWD, XSETUN, XSETF, IXSAV, IUMACH -- which handle error
messages from the solvers.

These routines are slated for elimination and replacement with a 
more intuitive interface.

This set is in fact a reduced version (sufficient for the needs of
ODEPACK) of a much larger error handling package from the SLATEC Library.
If your computer system already has the full SLATEC error handler, the
version provided here can be ignored.  If the reduced version is used,
its machine-dependent features should be checked first; see comments in
Subroutine XERRWD.

##### Non-standard code

4. ODEPACK contains a few instances where the Fortran Standard is violated:

   (a) In various places in the LSODES and LSODIS solvers, a call to a
   subroutine has a subscripted real array as an argument where the
   subroutine called expects an integer array.  Calls of this form
   occur in Subroutine DLSODES (to DSTODE), in DIPREP (to DPREP),
   in Subroutine DLSODIS (to DSTODI), and in DIPREPI (to DPREPI).
   Another such call occurs in the DLSODES demonstration program,
   from the main program to Subroutine SSOUT.  

   This is done in order
   to use work space in an efficient manner, as the same space is
   sometimes used for real work space and sometimes for integer work
   space.  If your compiler does not accept this feature, one possible
   way to get the desired result is to compile the called routines
   and calling routines in separate jobs, and then combine the binary
   modules in an appropriate manner. 

   If this procedure is still not
   acceptable under your system, it will be necessary to radically
   alter the structure of the array RWORK within the LSODES or LSODIS
   solver package.  (See also Note 5 below.)

   (b) Each ODEPACK solver treats the arguments NEQ, Y, RTOL, and ATOL
   as arrays, even though the length may be only 1.  Moreover,
   except for Y, the usage instructions say that these arguments
   may be either arrays or scalars.  If your system does not allow
   such a mismatch, then the documentation of these arguments
   should be changed accordingly.

5. For maximum storage economy, the LSODES and LSODIS solvers make use
of the real to integer wordlength ratio.  This is assumed to be an
integer L such that if a real array R and an integer array M occupy
the same space in memory, R(1) having the same bit address as M(1),
then R(I) has the same address as M((I-1)*L+1).  This ratio L is
usually 2 for double precision, and this is the value used in the
double precision version supplied.  If this value is incorrect, it
needs to be changed in two places:

  (a) The integer LENRAT is DATA-loaded in Subroutines DLSODES and
      DLSODIS to this ratio, shortly below the prologue.

  (b) The integer LRATIO is DATA-loaded in Subroutine CDRV to this
      ratio, shortly below the prologue of that routine.

(See comments in both places.)  If the ratio is not an integer, use
the greatest integer not exceeding the ratio.

6. For installation of ODEPACK on a Cray computer, the source files
supplied include compiler directives for the CFT compiler.  These have
the form CDIR$ IVDEP and occur prior to certain loops that involve
subscript shifts (and would otherwise not be vectorized).  These
directives are (or should be) treated as comments by any other compiler.

7. On first obtaining ODEPACK, the demonstration programs should be
compiled and executed prior to any other use of the solvers.  In most
cases, these exercise all of the major method options in each solver,
and are self-checking.  (In the case of LSODPK and LSODKR, the
demonstration programs are not self-checking, and for LSODKR only one
major method option is used.)  In any case, the output can be compared
with the sample output supplied, which was generated from the double
precision version of ODEPACK on a 32-bit computer.  When comparing your
output with that supplied, differences of 10-20% in the final values of

docs/CHANGELOG.md
## odepack Changelog

The intent of this log is to keep everyone in the loop about what's new
in the odepack project. It is a curated, chronologically ordered list
of notifications of notable events such as bug fixes, new features,
and usage changes.

"Do unto others as you would have them do unto you", as they say. When I
find OS (Open Source) resources, I am hoping a lot of these boxes can be
checked ...
   - [x] git repository on WWW (github)
   - [x] annotated source files with an open license
   - [x] unit test
   - [x] make(1) build
   - [x] fpm(1) build
   - [x] user manual (on-line)
   - [x] man-page
   - [x] app program
   - [x] demo program for public procedures
   - [x] developer documents (ford(1))
   - [x] CI/CD(Continious Integration/Development) verification (github actions)
   - [ ] registered in fpm(1) repository

---
**2021-02-10**  John S. Urban  <https://github.com/urbanjost>

### :green_circle: ADD:
     initial release on github
---

<!--
### :orange_circle: DIFF:
       + renamed ADVICE(3f) to ALERT(3f)
### :green_circle: ADD:
       + advice(3f) was added to provide a standardized message format simply.
### :red_circle: FIX:
       + </bo> did not work on several terminal types, changed it to a more
         universally accepted value.
-->


src/M_odepack.f90
module M_odepack
!
! dcopy can be replaced with array syntax
! SPAG removes save attribute and type from f90 function declaration
! SPAG removes external attribute
! determine what can be private and public
! many equality tests for real values; could replace
! simplify checkpoint capability, maybe TRANSFER()
! look for SAVEs that should be parameters
! remaining GOTOs
! replace dnrm2 with intrinsic norm2(3f) ?
! in production probably would make matrix routines separate so could use alternate library of BLAS matrix procedures
! same arrays passed on single call all over the place
! NOTE: DLSODKR did not initialize JROOT and neither did lsodkr.f90 example
!
implicit none
integer,parameter :: dp=kind(0.0d0)

private
!----------------------------------------------------------------------------------------------------------------------------------!
! da1 routines
PUBLIC  :: DEWSET
PUBLIC  :: DINTDY
PUBLIC  :: DSRCAR
PUBLIC  :: DSRCKR
PUBLIC  :: DSRCMA
PUBLIC  :: DSRCMS
PUBLIC  :: DSRCOM
PUBLIC  :: DSRCPK
PUBLIC  :: DUMACH ! COMPUTE THE UNIT ROUNDOFF OF THE MACHINE.
PUBLIC  :: DVNORM 
!-PUBLIC  :: DSTODE

!-- maybe private once modularized
public :: dmnorm 
public :: dpjibt
public :: dprepj
public :: dprepji
public :: dsolss
public :: dprja
public :: dprjis
public :: dprjs
public :: dslsbt
public :: dsolsy
public :: adjlr
public :: cdrv
public :: cntnzu
public :: daigbt
public :: dainvg
public :: dcfode ! Set ODE integrator coefficients.
public :: diprep
public :: diprepi
public :: dlhin
public :: drchek
public :: dstoda
public :: dstodpk
public :: dstoka
public :: jgroup
public :: odrv
!-- see nothing in documentation about being public
private :: datp
private :: datv
private :: dbnorm 
private :: ddecbt
private :: dfnorm 
private :: dhefa
private :: dhels
private :: dheqr
private :: dhesl
private :: dorthog
private :: dpcg
private :: dpcgs
private :: dpkset
private :: droots
private :: dsetpk
private :: dsolbt
private :: dsolpk
private :: dspigmr
private :: dspiom
private :: dusol
private :: md
private :: mdi
private :: mdm
private :: mdp
private :: mdu
private :: nnfc
private :: nnsc
private :: nntc
private :: nroc
private :: nsfc
private :: sro
private :: dnrm2
!-private :: dstodi
!-private :: dprepi
!-private :: dainvgs
!-private :: dprep
!----------------------------------------------------------------------------------------------------------------------------------!
! matrix routines
PUBLIC  :: DAXPY
PUBLIC  :: XSETUN
PUBLIC  :: XSETF
PUBLIC  :: DSCAL
PUBLIC  :: DCOPY
PUBLIC  :: XERRWD
PUBLIC  :: DGEFA
PUBLIC  :: DGESL
! should any matrix procedures be private?
private :: idamax
private :: ixsav
private :: ddot
private :: dgbsl
private :: dgbfa
! main routines
PUBLIC :: DLSODIS
PUBLIC :: DLSODI
PUBLIC :: DLSOIBT
PUBLIC :: DLSODKR
PUBLIC :: DLSODPK
PUBLIC :: DLSODE
PUBLIC :: DLSODA
PUBLIC :: DLSODES
PUBLIC :: DLSODAR
!----------------------------------------------------------------------------------------------------------------------------------!
! state save and restore

type dls002
   real(kind=dp) :: stifr
   integer       :: newt, nsfi, nslj, njev 
end type dls002
type(dls002),public,save :: dls

type dlss01
   real(kind=dp) :: con0, conmin, ccmxj, psmall, rbig, seth 
   integer       :: iplost, iesp, istatc, iys, iba, ibian, ibjan, ibjgp, ipian, ipjan, ipjgp, ipigp, ipr, ipc, ipic, ipisp, iprsp
   integer       :: ipa, lenyh, lenyhm, lenwk, lreq, lrat, lrest, lwmin, moss, msbj, nslj, ngp, nlu, nnz, nsp, nzl, nzu
end type dlss01
type(dlss01),public,save :: dlss

type dlsa01
   real(kind=dp) :: tsw, cm1(12), cm2(5), pdest, pdlast, ratio, pdnorm 
   integer       :: insufr, insufi, ixpr, icount, irflag, jtyp, mused, mxordn, mxords
end type dlsa01
type(dlsa01),public,save :: dlsa

type dlsr01
   real(kind=dp) :: alpha, x2, t0, tlast, toutc 
   integer       :: lg0, lg1, lgx, imax, last, irfnd, itaskc, ngc, nge
end type dlsr01
type(dlsr01),public,save :: dlsr

type dlpk01
   real(kind=dp) :: delt, epcon, sqrtn, rsqrtn 
   integer       :: jpre, jacflg, locwp, lociwp, lsavx, kmp, maxl, mnewt, nni, nli, nps, ncfn, ncfl 
   end type dlpk01
type(dlpk01),public,save :: dlpk

type dls001
   real(kind=dp) :: conit, crate, el(13), elco(13,12), hold, rmax, tesco(3,12), ccmax, el0, h, hmin, hmxi, hu, rc, tn, uround 
   integer       :: init,   mxstep,  mxhnil,  nhnil,  nslast,  nyh,   ialth,   ipup,    lmax,    meo
   integer       :: nqnyh,  nslp,    icf,     ierpj,  iersl,   jcur,  jstart,  kflag,   l,       lyh
   integer       :: lewt,   lacor,   lsavf,   lwm,    liwm,    meth,  miter,   maxord,  maxcor,  msbp
   integer       :: mxncf,  n,       nq,      nst,    nfe,     nje,   nqu
end type dls001
type(dls001),public,save :: dls1

private :: return_dls1_real
private :: return_dls1_int
private :: set_dls1_real
private :: set_dls1_int

contains
!----------------------------------------------------------------------------------------------------------------------------------!
!()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()!
!----------------------------------------------------------------------------------------------------------------------------------!

function return_dls1_real()
real(kind=dp),allocatable :: return_dls1_real(:)

   return_dls1_real=[ dls1%conit, dls1%crate, dls1%el(:), dls1%elco(:,:), dls1%hold, &
                    & dls1%rmax, dls1%tesco(:,:), dls1%ccmax, dls1%el0, dls1%h, dls1%hmin, &
                    & dls1%hmxi, dls1%hu, dls1%rc, dls1%tn, dls1%uround ]

end function return_dls1_real

function return_dls1_int()
integer,allocatable :: return_dls1_int(:)

   return_dls1_int=[ dls1%init, dls1%mxstep, dls1%mxhnil, dls1%nhnil, dls1%nslast, dls1%nyh, dls1%ialth, &
                   & dls1%ipup, dls1%lmax, dls1%meo, dls1%nqnyh, dls1%nslp, dls1%icf, dls1%ierpj, &
                   & dls1%iersl, dls1%jcur, dls1%jstart, dls1%kflag, dls1%l, dls1%lyh, dls1%lewt, &
                   & dls1%lacor, dls1%lsavf, dls1%lwm, dls1%liwm, dls1%meth, dls1%miter, dls1%maxord, &
                   & dls1%maxcor, dls1%msbp, dls1%mxncf, dls1%n, dls1%nq, dls1%nst, dls1%nfe, &
                   & dls1%nje, dls1%nqu ]

end function return_dls1_int

subroutine set_dls1_real(r)
real(kind=dp),intent(in) :: r(218)
   dls1%conit        =  r(1)
   dls1%crate        =  r(2)
   dls1%el           =  r(3:15)
   dls1%elco         =  reshape(r(16:16+156-1),[13,12])
   dls1%hold         =  r(172)
   dls1%rmax         =  r(173)
   dls1%tesco        =  reshape(r(174:174+36-1),[3,12])
   dls1%ccmax        =  r(210)
   dls1%el0          =  r(211)
   dls1%h            =  r(212)
   dls1%hmin         =  r(213)
   dls1%hmxi         =  r(214)
   dls1%hu           =  r(215)
   dls1%rc           =  r(216)
   dls1%tn           =  r(217)
   dls1%uround       =  r(218)

end subroutine set_dls1_real

subroutine set_dls1_int(int)
integer,intent(in) :: int(37)
   dls1%init    =  int(1)
   dls1%mxstep  =  int(2)
   dls1%mxhnil  =  int(3)
   dls1%nhnil   =  int(4)
   dls1%nslast  =  int(5)
   dls1%nyh     =  int(6)
   dls1%ialth   =  int(7)
   dls1%ipup    =  int(8)
   dls1%lmax    =  int(9)
   dls1%meo     =  int(10)
   dls1%nqnyh   =  int(11)
   dls1%nslp    =  int(12)
   dls1%icf     =  int(13)
   dls1%ierpj   =  int(14)
   dls1%iersl   =  int(15)
   dls1%jcur    =  int(16)
   dls1%jstart  =  int(17)
   dls1%kflag   =  int(18)
   dls1%l       =  int(19)
   dls1%lyh     =  int(20)
   dls1%lewt    =  int(21)
   dls1%lacor   =  int(22)
   dls1%lsavf   =  int(23)
   dls1%lwm     =  int(24)
   dls1%liwm    =  int(25)
   dls1%meth    =  int(26)
   dls1%miter   =  int(27)
   dls1%maxord  =  int(28)
   dls1%maxcor  =  int(29)
   dls1%msbp    =  int(30)
   dls1%mxncf   =  int(31)
   dls1%n       =  int(32)
   dls1%nq      =  int(33)
   dls1%nst     =  int(34)
   dls1%nfe     =  int(35)
   dls1%nje     =  int(36)
   dls1%nqu     =  int(37)

end subroutine set_dls1_int
!----------------------------------------------------------------------------------------------------------------------------------!
!()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()!
!----------------------------------------------------------------------------------------------------------------------------------!
! da1 routines
include "M_da1/dumach.inc"
include "M_da1/dcfode.inc"
include "M_da1/dmnorm.inc"
include "M_da1/dfnorm.inc"
include "M_da1/dbnorm.inc"
include "M_da1/dvnorm.inc"
include "M_da1/sro.inc"
include "M_da1/adjlr.inc"
include "M_da1/dewset.inc"
include "M_da1/jgroup.inc"
include "M_da1/odrv.inc"
include "M_da1/dhels.inc"
include "M_da1/cntnzu.inc"
include "M_da1/daigbt.inc"
include "M_da1/ddecbt.inc"
include "M_da1/dslsbt.inc"
include "M_da1/dsolbt.inc"
include "M_da1/md.inc"
include "M_da1/mdi.inc"
include "M_da1/mdm.inc"
include "M_da1/mdp.inc"
include "M_da1/mdu.inc"
include "M_da1/nnsc.inc"
include "M_da1/nntc.inc"
include "M_da1/cdrv.inc"
include "M_da1/dainvg.inc"
include "M_da1/dheqr.inc"
include "M_da1/dhesl.inc"
include "M_da1/dorthog.inc"
include "M_da1/dusol.inc"
include "M_da1/dhefa.inc"
include "M_da1/dpcg.inc"
include "M_da1/dspiom.inc"
include "M_da1/dpcgs.inc"
include "M_da1/nsfc.inc"
include "M_da1/dspigmr.inc"
include "M_da1/nroc.inc"
include "M_da1/nnfc.inc"
include "M_da1/dintdy.inc"
include "M_da1/droots.inc"
include "M_da1/dsolss.inc"
include "M_da1/dsolsy.inc"
include "M_da1/dsrcar.inc"
include "M_da1/dsrckr.inc"
include "M_da1/dsrcma.inc"
include "M_da1/dsrcms.inc"
include "M_da1/dsrcom.inc"
include "M_da1/dsrcpk.inc"
include "M_da1/dlhin.inc"
include "M_da1/diprep.inc"
include "M_da1/diprepi.inc"
include "M_da1/dprepji.inc"
include "M_da1/drchek.inc"
include "M_da1/dsetpk.inc"
include "M_da1/datv.inc"
include "M_da1/dprjis.inc"
include "M_da1/dprja.inc"
include "M_da1/dpjibt.inc"
include "M_da1/dpkset.inc"
include "M_da1/dsolpk.inc"
include "M_da1/dprjs.inc"
include "M_da1/datp.inc"
include "M_da1/dprepj.inc"
include "M_da1/dstoda.inc"
include "M_da1/dstoka.inc"
include "M_da1/dstodpk.inc"
!----------------------------------------------------------------------------------------------------------------------------------!
!()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()!
!----------------------------------------------------------------------------------------------------------------------------------!
! main routines
include "M_main/dlsodis.inc"
include "M_main/dlsodi.inc"
include "M_main/dlsoibt.inc"
include "M_main/dlsodkr.inc"
include "M_main/dlsodpk.inc"
include "M_main/dlsode.inc"
include "M_main/dlsoda.inc"
include "M_main/dlsodes.inc"
include "M_main/dlsodar.inc"
!----------------------------------------------------------------------------------------------------------------------------------!
!()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()!
!----------------------------------------------------------------------------------------------------------------------------------!
! matrix routines
include "M_matrix/daxpy.inc"
include "M_matrix/xsetun.inc"
include "M_matrix/xsetf.inc"
include "M_matrix/dscal.inc"
include "M_matrix/idamax.inc"
include "M_matrix/dcopy.inc"
include "M_matrix/ixsav.inc"
include "M_matrix/ddot.inc"
include "M_matrix/xerrwd.inc"
include "M_matrix/dgefa.inc"
include "M_matrix/dgesl.inc"
include "M_matrix/dgbsl.inc"
include "M_matrix/dgbfa.inc"
include "M_matrix/dnrm2.inc"
!----------------------------------------------------------------------------------------------------------------------------------!
!()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()!
!----------------------------------------------------------------------------------------------------------------------------------!
!-include "M_da1/dprep.inc"
!-include "M_da1/dainvgs.inc"
!-include "M_da1/dprepi.inc"
!-include "M_da1/dstodi.inc"
!-include "M_da1_/dstode.inc"
!----------------------------------------------------------------------------------------------------------------------------------!
!()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()!
!----------------------------------------------------------------------------------------------------------------------------------!
end module M_odepack
!----------------------------------------------------------------------------------------------------------------------------------!
!()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()()!
!----------------------------------------------------------------------------------------------------------------------------------!
!- TYPE MISMATCH
include "M_da1/dprep.inc"
include "M_da1/dainvgs.inc"
include "M_da1/dprepi.inc"
include "M_da1/dstodi.inc"
include "M_da1/dstode.inc"
